server/lomas_server/dp_queries/dp_libraries/smartnoise_synth.py

```diff
@@ -0,0 +1,307 @@
+                ) from e
+        return private_data
+
+    def _get_fit_model(
+        self,
+        private_data: pd.DataFrame,
+        transformer: TableTransformer,
+        query_json: dict,
+    ) -> Synthesizer:
+        """
+        Create and fit the synthesizer model.
+
+        Args:
+            private_data (pd.DataFrame): Private data for fitting the model
+            transformer (TableTransformer): Transformer to pre/postprocess data
+            query_json (SmartnoiseSynthModelCost): JSON request object for the query
+                synth_name (str): name of the Yanthesizer model to use
+                epsilon (float): epsilon budget value
+                delta (float): delta budget value
+                nullable (bool): True if some data cells may be null
+                synth_params (dict): Keyword arguments to pass to the synthesizer
+                    constructor.
+
+        Returns:
+            Synthesizer: Fitted synthesizer model
+        """
+        if query_json.delta:  # not all model take delta as argument
+            query_json.synth_params["delta"] = query_json.delta
+        if query_json.synth_name == SSynthSynthesizer.DP_CTGAN:
+            query_json.synth_params["disabled_dp"] = False
+
+        model = Synthesizer.create(
+            synth=query_json.synth_name,
+            epsilon=query_json.epsilon,
+            verbose=True,
+            **query_json.synth_params,
+        )
+        try:
+            model.fit(
+                data=private_data,
+                transformer=transformer,
+                preprocessor_eps=0.0,  # will error if not 0.0
+                nullable=query_json.nullable,
+            )
+        except Exception as e:
+            raise ExternalLibraryException(
+                DPLibraries.SMARTNOISE_SYNTH, "Error fitting model:" + str(e)
+            ) from e
+
+        return model
+
+    def query(self, query_json: SmartnoiseSynthModel) -> pd.DataFrame:
+        """Perform the query and return the response.
+
+        Args:
+            query_json (SmartnoiseSynthModel): The JSON request object for the query.
+
+        Raises:
+            ExternalLibraryException: For exceptions from libraries
+                external to this package.
+            InvalidQueryException: If the budget values are too small to
+                perform the query.
+
+        Returns:
+            pd.DataFrame: The resulting pd.DataFrame samples.
+        """
+        # Preprocessing information from metadata
+        metadata = self.private_dataset.get_metadata()
+        private_data = self.private_dataset.get_pandas_df()
+        private_data = self._preprocess_data(private_data, query_json)
+        transformer = self._prepare_data_transformer(
+            metadata, private_data, query_json
+        )
+
+        # Create and fit synthesizer
+        model = self._get_fit_model(private_data, transformer, query_json)
+        return serialise_model(model)
```

Comment:
Is it always safe to return the model instead of generated synthetic data? ie. does the model keep private data in its internal state (I would assume not though)?

Reply:
i think so. but i am also going to give the option to send samples back instead